### PR TITLE
Added 'bulk delete/edit' mode.

### PR DESCRIPTION
**Pull request by Slessi**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/.github/CONTRIBUTING.md). Confirmation: 💪😎👊

## What's in this pull request?

All of #1593 plus some of my own changes, implementing the "Bulk edit" mode.

Notable change: I delegated the selected messages list to the user as tracking with indexPaths doesn't play well with load earlier messages (index paths will change)

@ignacio-chiazzo Here is PR of my fork, rebased over develop. There were quite a few conflicts but I believe I've managed to resolve them correctly, tests still pass and demo project works with edit mode. I'm relatively new to Obj-C though so don't take my word for it :)

### CODE REVIEW

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 1]
```diff
@@ -732,6 +769,43 @@ - (void)collectionView:(JSQMessagesCollectionView *)collectionView
...
 
+#pragma mark - bulk edit mode support
+
+- (CGFloat)collectionView:(JSQMessagesCollectionView *)collectionView
+                   layout:(JSQMessagesCollectionViewFlowLayout *)collectionViewLayout editingOffsetForCellAtIndexPath:(NSIndexPath *)indexPath
+{
+    id<JSQMessageData> messageData = [collectionView.dataSource collectionView:collectionView messageDataForItemAtIndexPath:indexPath];
+
+    // outgoing messages aren't offsetted
+    return [self isOutgoingMessage:messageData] ? 0.0f : 30.0f;
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: Just a quick question, Why 30.0f?

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m [hunk 2]
```diff
@@ -83,6 +91,10 @@ - (void)jsq_configureFlowLayout
     _springinessEnabled = NO;
     _springResistanceFactor = 1000;
     
+    _transientEditingOverlayIndexPaths = @{kJSQCollectionUpdateTypeInsert : [NSMutableArray array],
+                                           kJSQCollectionUpdateTypeDelete : [NSMutableArray array]};
+
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: extra line here

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m [hunk 3]
```diff
@@ -218,6 +245,16 @@ - (NSMutableSet *)visibleIndexPaths
...
 }
 
+-(NSMutableSet *)editableIndexPaths
+{
+    if (!_editableIndexPaths) {
+        _editableIndexPaths = [NSMutableSet new];
+    }
+    return _editableIndexPaths;
+}
+
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: some extra lines here

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m [hunk 4]
```diff
@@ -355,6 +428,7 @@ - (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds
 - (void)prepareForCollectionViewUpdates:(NSArray *)updateItems
 {
     [super prepareForCollectionViewUpdates:updateItems];
+
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: extra line here

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m [hunk 5]
```diff
@@ -382,14 +456,82 @@ - (void)prepareForCollectionViewUpdates:(NSArray *)updateItems
...
+    }
 }
 
+
+
+
+- (void)prepareForCollectionViewUpdatesInEditMode:(NSArray *)updateItems{
+    
+    [updateItems enumerateObjectsUsingBlock:^(UICollectionViewUpdateItem *updateItem, NSUInteger index, BOOL *stop) {
+        if (updateItem.updateAction == UICollectionUpdateActionInsert) {
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: Maybe with a swift cases over .updateAction looks better

Comment (Slessi) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: Haven't understood this 1

Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]:
```
switch (.updateAction)
 {
     case UICollectionUpdateActionInsert:
              //piece of code
              break;
     case UICollectionUpdateActionDelete:
              //piece of code
              break;
     case UICollectionUpdateActionMove:
              //piece of code
              break;
}
```

**Comment by ignacio-chiazzo** [comment]
Another question : Why didn't you use the default icons (edit mode)?

**Comment by Slessi** [comment]
When I had made this fork it was to use it in my own project, and I didn't like the icons being used so I changed it to allow me to override them

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m (hunk 1 above)
Comment (Slessi) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: ¯\_(ツ)_/¯ this is one of the original guy's lines, don't know how he came to this value

**Comment by Slessi** [comment]
Removed extra lines & added the switch

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m [hunk 6]
```diff
@@ -382,14 +452,81 @@ - (void)prepareForCollectionViewUpdates:(NSArray *)updateItems
             }
         }
     }];
+    
+    if(self.editing) {
+        [self prepareForCollectionViewUpdatesInEditMode:updateItems];
+    }
 }
 
+
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: extra lines

File: JSQMessagesViewController/Views/JSQMessagesCollectionView.m [hunk 7]
```diff
@@ -194,4 +211,57 @@ - (void)messagesCollectionViewCellDidTapAccessoryButton:(JSQMessagesCollectionVi
...
+    [self.delegate collectionView:self editingOverlayAtIndexPath:indexPath becomeSelected:NO];
+}
+
+- (void)layoutWillMoveItemAtIndexPath:(NSIndexPath *)indexPath toIndexPath:(NSIndexPath *)newIndexPath
+{
+    if (indexPath == nil || newIndexPath == nil) {
+        return;
+    }
+
+    //this won't cover 100% of cases (like when user rearranging selected items)
```
Comment (ignacio-chiazzo) [inline JSQMessagesViewController/Views/JSQMessagesCollectionView.m]: We need to take care of this

**Comment by ignacio-chiazzo** [comment]
Would be good to make the "editingInactiveImage" and  "editingActiveImage" methods public so that the developer can customize the image, and as default, the application should take the image default.

**Comment by ignacio-chiazzo** [comment]
Could you make a fork of this branch in your repo?

**Comment by Slessi** [comment]
The following methods are in `JSQMessagesViewController` implementing default values, and can be overridden with a subclass -- Is this not sufficient?

```
- (UIImage *)editingActiveImage;
- (UIImage *)editingInactiveImage;
- (UIColor *)editingActiveColor;
- (UIColor *)editingInactiveColor;
```

File: JSQMessagesViewController/Views/JSQMessagesCollectionView.m (hunk 7 above)
Comment (Slessi) [inline JSQMessagesViewController/Views/JSQMessagesCollectionView.m]: Hmm, how would user rearrange selected items? The only time I ran into a problem with selected index paths was when using load earlier messages, which I resolved with my change to use `isEditableItemSelectedAtIndexPath:`, maybe this covered it?

**Comment by Terens777** [comment]
How to add this functionality to your project, your help is very necessary ?

**Comment by jessesquires** [comment]
Hello everyone!

I'm sorry to inform the community that I'm officially deprecating this project. 😢 Please read my blog post for details:

http://www.jessesquires.com/blog/officially-deprecating-jsqmessagesviewcontroller/

Thus, I'm closing all issues and pull requests and making the necessary updates to formally deprecate the library. I'm sorry if this is unexpected or disappointing. Please know that this was an extremely difficult decision to make. I'd like to thank everyone here for contributing and making this project so great. It was a fun 4 years. 😊 

Thanks for understanding,
— jsq

**Pull request closed without merging**